src/cryptonote_core/blockchain.cpp

```diff
@@ -5531,6 +5571,33 @@ void Blockchain::cache_block_template(const block &b, const cryptonote::account_
   m_btc_valid = true;
 }
 
+void Blockchain::send_miner_notifications(const crypto::hash &prev_id, uint64_t already_generated_coins)
+{
+  if (m_miner_notifiers.empty())
+    return;
+
+  const uint64_t height = m_db->height();
```

Comment:
Would be nice if some of these values could be retrieved lazily only when there is a sub to the corresponding endpoint. But this notification is infrequent, so I guess not worth butchering the interface -

Could you add a comment here about lazily retrieving though? Just to document somewhere that the data (usually) gets dropped on the floor. The notifier signature could be updated to `boost::optional<miner_data>(const crypto::hash&, std::uint64_t /* already_generated_coins */, blockchain&, const miner_data*)`, which is a bit more efficient but admittedly funky.

Basically the first notifier that has a subscriber retrieves the data and returns it for the next one, etc.

Reply:
`m_miner_notifiers` is not empty only when there is `--zmq-pub` in monerod command line, so it early outs in 99% cases. The remaining monerod setups that do use `--zmq-pub` most likely will use it for p2pool, so they will use this data. Refactoring it makes sense later when there are more uses for ZMQ.

Reply:
Mining is not the only use case for pub/sub, so there will be cases where all of this information is collected only to be dropped. The other endpoints are forwarding data that is already available.

Reply:
Is there an easy way to detect if anyone is connected to zmq-pub port and subscribed to `json-full-miner_data`? I tried to do something like this initially, but I couldn't find a way to detect connected subscribers reliably. IIRC ZMQ somehow cached it internally and `add_subscriptions` was only called once even when I connected two subscribers, or something like that.

Reply:
The `miner_subs_` array should be an accurate count. ZMQ may coalesce TCP connections on the client side, so you need to create two `zmq_context` objects or use a separate process entirely. If this still doesn't work, then I'd have to re-read the documentation because I misunderstood how `XPUB` works.